### DISCUSSION TITLE
Cite URL - URL and accessed by missing, any ideas?

### DISCUSSION THREAD

**Discussion by HachimanSec** [retorquere/zotero-better-bibtex#1995] [Q&A] [answered]
Hi everyone, 

I am new to LaTeX and I havent found anything useful on the net, so I come here and hope you have a hint for me. 

Right now I am using Zotero and Better Bibtex for my thesis (awesome work with Better Bibtex BTW, thanks a lot for that 👍). 
LaTeX writing happens in overleafe.com

This is the bibtex entry: 

```
@online{PrivacyShieldProgram,
  title = {Privacy {{Shield Program Overview}} | {{Privacy Shield}}},
  url = {https://www.privacyshield.gov/Program-Overview},
  urldate = {2021-11-27},
  langid = {english},
}
```

It only appears like this in the bibliography (no URL, no accessed by date):
**[19] Privacy Shield Program Overview | Privacy Shield.**

According to Overleafs own documentation "How to Write a Thesis in LaTeX (Part 4): Bibliographies with BibLaTeX" https://ogy.de/uel5, this is exactly how it should work. 

Any ideas? Your help is highly appreciated! 

Regards
Tom

**Comment by retorquere** [comment]
I'd love to help you, but I think you'll get the most informative answers on tex.stackexchange.com

**Comment by HachimanSec** [comment]
Thx @retorquere!

I eventually found the issue (a quite simple one). I will leave my comments here, maybe they are helpful to someone else.  

If the university forces you to use a style which doesnt support it, a workaround can be to add note in your bibtex. 
Something the Overleaf team suggested:

> Alternatively you can add an extra field:
> note = {Accessed 2021-11-27}
> 
> All bibliography styles should support and print the "note" field.

As mentioned below, I am trying to achieve this now with a postscript as suggested by retorquere.

**Reply by retorquere** [reply]
You can automate this with a postscript if so desired.

**Reply by HachimanSec** [reply]
Thank you for the hint. I just found out that \bibliographystyle{unsrtnat} doesnt work. 

I am not trying your suggested workaround @retorquere. I found this link on your site: https://retorque.re/zotero-better-bibtex/exporting/scripting/#add-accessdate-url-for-bibtex

I used the export now to force a new export. Unfortunately it doesnt work. Any idea what could be wrong? 

The code I use (I used itemType webpage and Web Page, none works):
```
if (Translator.BetterBibTeX && item.itemType === 'Web Page') {
    if (item.accessDate) {
      reference.add({ name: 'note', value: "(accessed " + item.accessDate.replace(/\s*T?\d+:\d+:\d+.*/, '') + ")" });
    }
    if (item.url) {
      reference.add({ name: 'howpublished', bibtex: "{\\url{" + reference.enc_verbatim({value: item.url}) + "}}" });
    }
  }
```

**Reply by retorquere** [reply]
You'd have to get me a support log for that entry for me to be able to tell.

**Reply by retorquere** [reply]
Oh one thing beforehand - the itemType would be `'webPage'`. If you want to know what the data available to the postscript looks like, export the item as `BetterBibTeX JSON`.

**Reply by HachimanSec** [reply]
Thank you for the offer. The ID is STVFL98H-refs-euc

I tried it with webPage as you suggested, and also with webpage (as mentioned in the json). However the notes section is always empty.

**Reply by retorquere** [reply]
Ah sorry, it was indeed `webpage`, not `webPage`. The debug ID gives me a copy of the item, so I don't need it inline. With `webpage`, I get

```
@misc{PrivacyShieldProgram,
  title = {Privacy Shield Program Overview | Privacy Shield},
  langid = {english},
  note = {(accessed 2021-11-27)},
  howpublished = {\url{https://www.privacyshield.gov/Program-Overview}}
}
```

**Reply by HachimanSec** [reply]
Thanks. Just tried it again, unfortunately still no success. 

I modified the postcript 
```
if (Translator.BetterBibTeX && item.itemType === 'webpage') {
    if (item.accessDate) {
      reference.add({ name: 'note', value: "(accessed " + item.accessDate.replace(/\s*T?\d+:\d+:\d+.*/, '') + ")" });
    }
    if (item.url) {
      reference.add({ name: 'howpublished', bibtex: "{\\url{" + reference.enc_verbatim({value: item.url}) + "}}" });
    }
  }
```
But still all I get is the item without note. 

```
@online{EDOBStellungnahmeSchrems,
  title = {{{EDÖB}} - {{Stellungnahme}} Zu {{Schrems II}} - Daten:Recht – Das {{Datenschutz-Team}} von {{Walder Wyss}}},
  url = {https://datenrecht.ch/edoeb-stellungnahme-zu-schrems-ii/},
  urldate = {2021-11-27},
  file = {/Users/tom/Zotero/storage/WZNFAH6H/edoeb-stellungnahme-zu-schrems-ii.html}
}
```

Any idea what I could check?

**Reply by HachimanSec** [reply]
EDIT:
My bad. I was too dumb and checked the wrong export. The original collection was BetterBibLaTeX, instead of BetterBibTex. Obviously the if clause never went true. 
After adjusting it it works flawlessly now!

Thanks again for your help and work!

```
if (Translator.BetterBibLaTeX && item.itemType === 'webpage') {
    if (item.accessDate) {
      reference.add({ name: 'note', value: "(accessed " + item.accessDate.replace(/\s*T?\d+:\d+:\d+.*/, '') + ")" });
    }
    if (item.url) {
      reference.add({ name: 'howpublished', bibtex: "{\\url{" + reference.enc_verbatim({value: item.url}) + "}}" });
    }
  }
```

---------
OLD, see above!
Interesting thing I noticed (BTW, it works now). 

I created a new collection at the top level. Moved a few items there and created a new auto export. Then all of a sudden it worked.
Not sure why but maybe it gives you some feedback.

The samples now look like expected:
```
@misc{SR235Bundesgesetz,
  title = {{{SR}} 235.1 - {{Bundesgesetz}} Vom 19. {{Juni}} 1992 \"Uber Den {{Datenschutz}} ({{DSG}})},
  note = {(accessed 2021-12-03)},
  howpublished = {\url{https://www.fedlex.admin.ch/eli/cc/1993/1945_1945_1945/de}}
}

```

**Reply by retorquere** [reply]
Bummer. That means it's most likely a cache issue, but changing the postscript (or any other preference) should have dropped the cache. But if you no longer experience the problem, we can't diagnose.

**Comment by HachimanSec** [comment, marked as answer]
As mentioned above, I used Translator.BetterBibLaTeX and it works now:

```
if (Translator.BetterBibLaTeX && item.itemType === 'webpage') {
    if (item.accessDate) {
      reference.add({ name: 'note', value: "(accessed " + item.accessDate.replace(/\s*T?\d+:\d+:\d+.*/, '') + ")" });
    }
    if (item.url) {
      reference.add({ name: 'howpublished', bibtex: "{\\url{" + reference.enc_verbatim({value: item.url}) + "}}" });
    }
  }
```
